[AutoFill Debugging] Part 3: Add support for new text extraction item types and metadata

whsieh opened the pull request:
#### 20c66472f87c18ac32c6b2dc06b789c80b8b0ca1
<pre>
[AutoFill Debugging] Part 3: Add support for new text extraction item types and metadata
<a href="https://bugs.webkit.org/show_bug.cgi?id=297419">https://bugs.webkit.org/show_bug.cgi?id=297419</a>
<a href="https://rdar.apple.com/158352832">rdar://158352832</a>

Reviewed by Abrar Rahman Protyasha.

Add support for new text extraction item types:

• LinkItemData
• ContentEditableData
• TextFormControlData
• SelectData

...as well as additional metadata per item:

• `role` DOM attribute
• Various `aria-*` DOM attributes
• Event listeners
• Unique node IDs (only if needed)

There should be no change in behavior yet, since these items aren&apos;t surfaced through any WebKit API.

* LayoutTests/fast/text-extraction/avoid-extracting-password-fields.html:
* LayoutTests/fast/text-extraction/basic-text-extraction.html:
* LayoutTests/fast/text-extraction/text-extraction-display-contents.html:
* LayoutTests/fast/text-extraction/text-extraction-visible-overflow-expected.html:
* LayoutTests/fast/text-extraction/text-extraction-visible-overflow.html:
* LayoutTests/fast/text-extraction/text-extraction-when-scrolled.html:

Adjust existing tests to continue exercising existing text extraction behaviors by setting the
`mergeParagraphs` and `skipNearlyTransparentContent` configuration flags.

* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractItemData):
(WebCore::TextExtraction::shouldIncludeNodeIdentifier):
(WebCore::TextExtraction::extractRecursive):
(WebCore::TextExtraction::extractItem):
* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
(WKTextExtractionEventListenerTypes.description):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::containerType):
(WebKit::eventListenerTypes):
(WebKit::createItemWithChildren):
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKWebView+TextExtraction.swift:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestTextExtraction):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestTextExtraction):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
* Tools/TestRunnerShared/UIScriptContext/UIScriptControllerShared.cpp:
(WTR::toTextExtractionOptions):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::requestTextExtraction):

Canonical link: <a href="https://commits.webkit.org/298726@main">https://commits.webkit.org/298726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/32adb7fd1732849787308ca06bdb1e26cb119c37

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116468 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36129 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26680 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122525 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67031 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118357 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36823 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44717 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/88445 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119417 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29361 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104486 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68888 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28432 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22593 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66206 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98736 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22753 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125674 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43362 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32561 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97151 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43727 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100695 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96946 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24674 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42238 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20154 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43250 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48841 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42716 "Built successfully") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/46056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44421 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -270,15 +281,20 @@ static bool shouldTreatAsPasswordField(const Element* element)
+    if (!renderer)
+        return { SkipExtraction::SelfAndSubtree };
+
+    if (context.ignoreTransparency && renderer->style().opacity() < minOpacityToConsiderVisible)
```
Hmmm, we want to ignore transparency but then we consult the renderer opacity?

whsieh replied:
Oh, good point — the name is a bit confusing, since "ignore" in this context means "don't include fully/nearly transparent elements in the extraction results". It should really be called something like `skipNearlyTransparentContent` instead.

aprotyas commented on Source/WebCore/page/text-extraction/TextExtractionTypes.h:
```diff
@@ -39,12 +40,14 @@ enum class EventListenerCategory : uint16_t {
-    Focus       = 1 << 7,
-    Form        = 1 << 8,
-    Media       = 1 << 9,
+    Keyboard    = 1 << 4,
```
EventListenerCategory can now use a `uint8_t` backing type

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -6388,8 +6391,18 @@ - (void)_requestTextExtraction:(_WKTextExtractionConfiguration *)configuration c
+        if (!strongSelf)
+            return completionHandler(nil);
+
+        RetainPtr rootItem = WebKit::createItem(item, [strongSelf](auto& rectInRootView) -> WebCore::FloatRect {
```
```suggestion
        RetainPtr rootItem = WebKit::createItem(std::forward<decltype(item)>(item), [strongSelf](auto& rectInRootView) -> WebCore::FloatRect {
```

aprotyas replied:
oh, not really important given we only take const lvalue refs anyway
```cpp
createItem(const WebCore::TextExtraction::Item&
```

whsieh replied:
Works for me — thanks for the review!

whsieh committed:
[AutoFill Debugging] Part 3: Add support for new text extraction item types and metadata
https://bugs.webkit.org/show_bug.cgi?id=297419
rdar://158352832

Reviewed by Abrar Rahman Protyasha.

Add support for new text extraction item types:

• LinkItemData
• ContentEditableData
• TextFormControlData
• SelectData

...as well as additional metadata per item:

• `role` DOM attribute
• Various `aria-*` DOM attributes
• Event listeners
• Unique node IDs (only if needed)

There should be no change in behavior yet, since these items aren't surfaced through any WebKit API.

* LayoutTests/fast/text-extraction/avoid-extracting-password-fields.html:
* LayoutTests/fast/text-extraction/basic-text-extraction.html:
* LayoutTests/fast/text-extraction/text-extraction-display-contents.html:
* LayoutTests/fast/text-extraction/text-extraction-visible-overflow-expected.html:
* LayoutTests/fast/text-extraction/text-extraction-visible-overflow.html:
* LayoutTests/fast/text-extraction/text-extraction-when-scrolled.html:

Adjust existing tests to continue exercising existing text extraction behaviors by setting the
`mergeParagraphs` and `skipNearlyTransparentContent` configuration flags.

* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractItemData):
(WebCore::TextExtraction::shouldIncludeNodeIdentifier):
(WebCore::TextExtraction::extractRecursive):
(WebCore::TextExtraction::extractItem):
* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
(WKTextExtractionEventListenerTypes.description):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::containerType):
(WebKit::eventListenerTypes):
(WebKit::createItemWithChildren):
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKWebView+TextExtraction.swift:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestTextExtraction):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestTextExtraction):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
* Tools/TestRunnerShared/UIScriptContext/UIScriptControllerShared.cpp:
(WTR::toTextExtractionOptions):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::requestTextExtraction):

Canonical link: https://commits.webkit.org/298726@main
